runtime/backtrace_byt.c

```diff
@@ -341,43 +341,50 @@ static void read_main_debug_info(struct debug_info *di)
   }
 
   fd = caml_attempt_open(&exec_name, &trail, 1);
-  if (fd < 0){
-    caml_fatal_error ("executable program file not found");
-    CAMLreturn0;
-  }
 
-  caml_read_section_descriptors(fd, &trail);
-  if (caml_seek_optional_section(fd, &trail, "DBUG") != -1) {
-    chan = caml_open_descriptor_in(fd);
-
-    num_events = caml_getword(chan);
-    events = caml_alloc(num_events, 0);
-
-    for (i = 0; i < num_events; i++) {
-      orig = caml_getword(chan);
-      evl = caml_input_val(chan);
-      caml_input_val(chan); /* Skip the list of absolute directory names */
-      /* Relocate events in event list */
-      for (l = evl; l != Val_int(0); l = Field(l, 1)) {
-        value ev = Field(l, 0);
-        Field(ev, EV_POS) = Val_long(Long_val(Field(ev, EV_POS)) + orig);
+  if (fd >= 0) {
+    caml_read_section_descriptors(fd, &trail);
+    if (caml_seek_optional_section(fd, &trail, "DBUG") != -1) {
+      chan = caml_open_descriptor_in(fd);
+
+      num_events = caml_getword(chan);
+      events = caml_alloc(num_events, 0);
+
+      for (i = 0; i < num_events; i++) {
+        orig = caml_getword(chan);
+        evl = caml_input_val(chan);
+        caml_input_val(chan); /* Skip the list of absolute directory names */
+        /* Relocate events in event list */
+        for (l = evl; l != Val_int(0); l = Field(l, 1)) {
+          value ev = Field(l, 0);
+          Field(ev, EV_POS) = Val_long(Long_val(Field(ev, EV_POS)) + orig);
+        }
+        /* Record event list */
+        Store_field(events, i, evl);
       }
-      /* Record event list */
-      Store_field(events, i, evl);
-    }
 
-    caml_close_channel(chan);
+      caml_close_channel(chan);
 
-    di->events = process_debug_events(caml_start_code, events, &di->num_events);
+      di->events =
+        process_debug_events(caml_start_code, events, &di->num_events);
+    } else {
+      close(fd);
+    }
+    fd = 0;
```

Comment:
With this write, isn't it the case that `fd` is always `0` at the `CAMLreturnT` point below? I don't understand when it is that we may have a non-zero return value.